### PR TITLE
fetch-kde-qt.sh: support current layout of download.kde.org

### PR DESCRIPTION
**Pull request by hamburger1984**
###### Motivation for this change

The script fetch-kde-qt.sh stopped working for the (new?) layout of download.kde.org. This PR restores functionality to fetch source package names and sha256 sums for plasma, kde and qt to write them into srcs.nix files.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Thra11** [comment]
Thanks for all your work on this!
When I try to use this to update kde-frameworks to 5.76:
```
./maintainers/scripts/fetch-kde-qt.sh pkgs/development/libraries/kde-frameworks/
```
it eventually fails with:
```
Saving to: ‘attica-5.76.0.tar.xz.mirrorlist’

attica-5.76.0.tar.xz.mirrorlist                    [ <=>                                                                                               ]  26.89K  --.-KB/s    in 0.08s   

2020-11-23 16:50:48 (357 KB/s) - ‘attica-5.76.0.tar.xz.mirrorlist’ saved [27537]

--2020-11-23 16:50:48--  https://download.kde.org/stable/frameworks/5.76/kxmlrpcclient-5.76.0.tar.xz.mirrorlist
Resolving download.kde.org (download.kde.org)... 168.119.32.158, 2a01:4f8:242:1ed5::3
Connecting to download.kde.org (download.kde.org)|168.119.32.158|:443... connected.
HTTP request sent, awaiting response... 404 Not Found
2020-11-23 16:50:48 ERROR 404: Not Found.
```
(plasma 20.03 needs a newer extra-cmake-modules)

File: maintainers/scripts/fetch-kde-qt.sh [hunk 1]
```diff
@@ -14,19 +14,48 @@ fi
...
+# for KDE:
+#  > wget will just find *.sig files - we'll fetch *.mirrorlist files manually and extract sha256 sums from the content
+# for QT:
+#  > wget will "traverse" *.mirrorlist files to find *.sha256 files
+# Keep *.sig files and *.sha256 files only
+
+# Delete *.mirrorlist (they were only needed to find *.sha256 files)
 find -type f -name '*.mirrorlist' -delete
 
+#read -p "See current files in $(pwd)"
```
Comment (SuperSandro2000) [inline maintainers/scripts/fetch-kde-qt.sh:27]: Is this from debugging?

File: maintainers/scripts/fetch-kde-qt.sh [hunk 2]
```diff
@@ -14,19 +14,48 @@ fi
...
+    sha256=$(gawk '{ print $1 }' "$src")
     nameVersion="${filename%.tar.*}"
     name=$(echo "$nameVersion" | sed -e 's,-[[:digit:]].*,,' | sed -e 's,-opensource-src$,,' | sed -e 's,-everywhere-src$,,')
     version=$(echo "$nameVersion" | sed -e 's,^\([[:alpha:]][[:alnum:]]*-\)\+,,')
-    echo "$name,$version,$src,$filename" >>$csv
+    path=$(dirname ${src:2})
+    echo "$name,$version,$sha256,$filename,$path" >>$csv
 done
 
+#read -p "See CSV in $csv"
```
Comment (SuperSandro2000) [inline maintainers/scripts/fetch-kde-qt.sh:57]: see above comment

File: maintainers/scripts/fetch-kde-qt.sh (hunk 1 above)
Comment (hamburger1984) [reply maintainers/scripts/fetch-kde-qt.sh:27]: yes - can be removed.

**Comment by hamburger1984** [comment]
> Thanks for all your work on this!
> When I try to use this to update kde-frameworks to 5.76:
> 
> ```
> ./maintainers/scripts/fetch-kde-qt.sh pkgs/development/libraries/kde-frameworks/
> ```
> 
> it eventually fails with:
> 
> ```
> Saving to: ‘attica-5.76.0.tar.xz.mirrorlist’
> 
> attica-5.76.0.tar.xz.mirrorlist                    [ <=>                                                                                               ]  26.89K  --.-KB/s    in 0.08s   
> 
> 2020-11-23 16:50:48 (357 KB/s) - ‘attica-5.76.0.tar.xz.mirrorlist’ saved [27537]
> 
> --2020-11-23 16:50:48--  https://download.kde.org/stable/frameworks/5.76/kxmlrpcclient-5.76.0.tar.xz.mirrorlist
> Resolving download.kde.org (download.kde.org)... 168.119.32.158, 2a01:4f8:242:1ed5::3
> Connecting to download.kde.org (download.kde.org)|168.119.32.158|:443... connected.
> HTTP request sent, awaiting response... 404 Not Found
> 2020-11-23 16:50:48 ERROR 404: Not Found.
> ```
> 

Thanks! Will take a look at this.

File: maintainers/scripts/fetch-kde-qt.sh (hunk 2 above)
Comment (hamburger1984) [reply maintainers/scripts/fetch-kde-qt.sh:57]: will remove it.

**Comment by hamburger1984** [comment]
@Thra11 should work for kxmlrpclient (and other packages in subfolders) as well now.

**Comment by Thra11** [comment]
Works for me!

**Comment by hamburger1984** [comment]
Should I open an issue for this to get a better chance of getting reviewed and/or merged?
@SuperSandro2000 (don't know if you're the right one to ask)

**Comment by SuperSandro2000** [comment]
> Should I open an issue for this to get a better chance of getting reviewed and/or merged?

Not required. Backlog is just to big.

**Comment by SuperSandro2000** [comment]
If you do not supply $1 you get the following:

```
./maintainers/scripts/fetch-kde-qt.sh
these paths will be fetched (0.63 MiB download, 3.09 MiB unpacked):
  /nix/store/xax35k4di9hxm0xsim0lbfdsami32dip-wget-1.20.3
copying path '/nix/store/xax35k4di9hxm0xsim0lbfdsami32dip-wget-1.20.3' from 'https://cache.nixos.org'...
./maintainers/scripts/fetch-kde-qt.sh: line 7: $1: unbound variable
```


Can you default this to ``--help`` and add an usage example there?

**Comment by SuperSandro2000** [comment]
> @SuperSandro2000 (don't know if you're the right one to ask)

I can definitely merge it and as it is not something bootstrap, stdenv or other very crucial parts I feel comfortable when it works.

But I found another bug:

```
FINISHED --2020-12-02 10:18:34--
Total wall clock time: 3m 37s
Downloaded: 101 files, 888K in 0.3s (3.41 MB/s)
--2020-12-02 10:18:34--  https://download.kde.org/kde/ftp/stable/frameworks/5.73/portingAids/kxmlrpcclient-5.73.0.tar.xz.mirrorlist
Resolving download.kde.org (download.kde.org)... 168.119.32.158, 2a01:4f8:242:1ed5::3
Connecting to download.kde.org (download.kde.org)|168.119.32.158|:443... connected.
HTTP request sent, awaiting response... 404 Not Found
2020-12-02 10:18:35 ERROR 404: Not Found
```

**Comment by hamburger1984** [comment]
Thanks for the feedback, @SuperSandro2000! I'll have a look and come back to you.

**Comment by hamburger1984** [comment]
> > @SuperSandro2000 (don't know if you're the right one to ask)
> 
> I can definitely merge it and as it is not something bootstrap, stdenv or other very crucial parts I feel comfortable when it works.
> 
> But I found another bug:
> 
> ```
> FINISHED --2020-12-02 10:18:34--
> Total wall clock time: 3m 37s
> Downloaded: 101 files, 888K in 0.3s (3.41 MB/s)
> --2020-12-02 10:18:34--  https://download.kde.org/kde/ftp/stable/frameworks/5.73/portingAids/kxmlrpcclient-5.73.0.tar.xz.mirrorlist
> Resolving download.kde.org (download.kde.org)... 168.119.32.158, 2a01:4f8:242:1ed5::3
> Connecting to download.kde.org (download.kde.org)|168.119.32.158|:443... connected.
> HTTP request sent, awaiting response... 404 Not Found
> 2020-12-02 10:18:35 ERROR 404: Not Found
> ```

How did you trigger this? What's the content of your fetch.sh in pkgs/development/libraries/kde-framework? That "ftp" in the path looks suspicious. It should be something like https://download.kde.org/stable/frameworks/5.73/

**Comment by hamburger1984** [comment]
> If you do not supply $1 you get the following:
> 
> ```
> ./maintainers/scripts/fetch-kde-qt.sh
> these paths will be fetched (0.63 MiB download, 3.09 MiB unpacked):
>   /nix/store/xax35k4di9hxm0xsim0lbfdsami32dip-wget-1.20.3
> copying path '/nix/store/xax35k4di9hxm0xsim0lbfdsami32dip-wget-1.20.3' from 'https://cache.nixos.org'...
> ./maintainers/scripts/fetch-kde-qt.sh: line 7: $1: unbound variable
> ```
> 
> Can you default this to `--help` and add an usage example there?
> 
> Something like eg `./maintainers/scripts/fetch-kde-qt.sh pkgs/development/libraries/kde-frameworks/` is totally fine.

Done.

**Comment by SuperSandro2000** [comment]
> How did you trigger this? 

I did ``nixpkgs-review pr 104684``, ``cd nixpkgs`` and run ``./maintainers/scripts/fetch-kde-qt.sh pkgs/development/libraries/kde-frameworks/``.

> What's the content of your fetch.sh in pkgs/development/libraries/kde-framework? 

```
WGET_ARGS=(https://download.kde.org/stable/frameworks/5.73/)
```

Could it be that I am using wget2?

**Comment by hamburger1984** [comment]
> > How did you trigger this?
> 
> I did `nixpkgs-review pr 104684`, `cd nixpkgs` and run `./maintainers/scripts/fetch-kde-qt.sh pkgs/development/libraries/kde-frameworks/`.
> 
> > What's the content of your fetch.sh in pkgs/development/libraries/kde-framework?
> 
> ```
> WGET_ARGS=(https://download.kde.org/stable/frameworks/5.73/)
> ```
> 
> Could it be that I am using wget2?

I did exactly the same and it works for me. My `wget --version` is 1.20.3. `which wget` is `/run/current-system/sw/bin/wget` which links to `/nix/store/s9mf59032b6nyd3bff3rp27zh439k7a4-wget-1.20.3/bin/wget`

**Comment by hamburger1984** [comment]
How do I upgrade to wget2? @SuperSandro2000

**Comment by SuperSandro2000** [comment]
> How do I upgrade to wget2? @SuperSandro2000

apt get install wget2 😂

**Comment by SuperSandro2000** [comment]
Without wget2 I get the following error:

```
Removing stable/frameworks/5.73/portingAids/index.html?C=S;O=D.tmp.tmp since it should be rejected.

--2020-12-03 10:37:34--  https://download.kde.org/stable/frameworks/5.73/portingAids/?C=D;O=D
Reusing existing connection to download.kde.org:443.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [text/html]
--2020-12-03 10:37:34--  https://download.kde.org/stable/frameworks/5.73/portingAids/?C=D;O=D
Reusing existing connection to download.kde.org:443.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [text/html]
Saving to: ‘stable/frameworks/5.73/portingAids/index.html?C=D;O=D.tmp.tmp’

stable/frameworks/5.73/portingAids/index.html?     [ <=>                                                                                                 ]  15.77K  --.-KB/s    in 0.002s

2020-12-03 10:37:34 (6.40 MB/s) - ‘stable/frameworks/5.73/portingAids/index.html?C=D;O=D.tmp.tmp’ saved [16147]

Removing stable/frameworks/5.73/portingAids/index.html?C=D;O=D.tmp.tmp since it should be rejected.

FINISHED --2020-12-03 10:37:34--
Total wall clock time: 32s
Downloaded: 101 files, 888K in 0.3s (2.85 MB/s)
--2020-12-03 10:37:34--  https://download.kde.org/pub/kde/stable/frameworks/5.73/portingAids/kross-5.73.0.tar.xz.mirrorlist
Resolving download.kde.org (download.kde.org)... 168.119.32.158, 2a01:4f8:242:1ed5::3
Connecting to download.kde.org (download.kde.org)|168.119.32.158|:443... connected.
HTTP request sent, awaiting response... 404 Not Found
2020-12-03 10:37:35 ERROR 404: Not Found
```

**Comment by hamburger1984** [comment]
> Without wget2 I get the following error:
> 
> ```
> Removing stable/frameworks/5.73/portingAids/index.html?C=S;O=D.tmp.tmp since it should be rejected.
> 
> --2020-12-03 10:37:34--  https://download.kde.org/stable/frameworks/5.73/portingAids/?C=D;O=D
> Reusing existing connection to download.kde.org:443.
> HTTP request sent, awaiting response... 200 OK
> Length: unspecified [text/html]
> --2020-12-03 10:37:34--  https://download.kde.org/stable/frameworks/5.73/portingAids/?C=D;O=D
> Reusing existing connection to download.kde.org:443.
> HTTP request sent, awaiting response... 200 OK
> Length: unspecified [text/html]
> Saving to: ‘stable/frameworks/5.73/portingAids/index.html?C=D;O=D.tmp.tmp’
> 
> stable/frameworks/5.73/portingAids/index.html?     [ <=>                                                                                                 ]  15.77K  --.-KB/s    in 0.002s
> 
> 2020-12-03 10:37:34 (6.40 MB/s) - ‘stable/frameworks/5.73/portingAids/index.html?C=D;O=D.tmp.tmp’ saved [16147]
> 
> Removing stable/frameworks/5.73/portingAids/index.html?C=D;O=D.tmp.tmp since it should be rejected.
> 
> FINISHED --2020-12-03 10:37:34--
> Total wall clock time: 32s
> Downloaded: 101 files, 888K in 0.3s (2.85 MB/s)
> --2020-12-03 10:37:34--  https://download.kde.org/pub/kde/stable/frameworks/5.73/portingAids/kross-5.73.0.tar.xz.mirrorlist
> Resolving download.kde.org (download.kde.org)... 168.119.32.158, 2a01:4f8:242:1ed5::3
> Connecting to download.kde.org (download.kde.org)|168.119.32.158|:443... connected.
> HTTP request sent, awaiting response... 404 Not Found
> 2020-12-03 10:37:35 ERROR 404: Not Found
> ```

uhm.. I don't understand what caused the script to look at https://download.kde.org/pub/kde/stable/frameworks/.. (404) while it was looking at https://download.kde.org/stable/frameworks/.. (200)

**Comment by hamburger1984** [comment]
We might as well close this - probably made obsolete by commit https://github.com/NixOS/nixpkgs/commit/3db04334e3eb57839b397c6d0de94d55dd18a812

**Pull request closed without merging**